### ISSUE TITLE
In method IMonitorInitialize.afterInitialize is not known result of initial propagation

### ISSUE THREAD

**Issue by gregy4** [chocoteam/choco-solver#581]
During my analysis of unusual results inside afterInitialize method I found that the method is called also in case when initial propagation ended because of contradiction. In that case domains of some variables can be not completely reduced since choco solver probably stops propagations on first contradiction.

Quick look at the code showed that result of initial propagation can be probably deduced by nonzero Measure.failCount but it is not nice solution. I recommend following variants
- add argument to afterInitializeMethod to inform about result of initial propagation
- update feasible variable so Solver.isFeasible method returns inside afterInitializeMethod valid value
- don't call afterInitializeMethod at all in this case since only part of propagators was executed 

Variants 1 and 3 are preferred by me.

Situation seems same in choco 3.3.3 and current version 4.0.6.

**Comment by cprudhom** [comment]
Hi, 

What is the underlying problem for you ? Because, I can not see any issue here.

I suppose you are referring to [this call](https://github.com/chocoteam/choco-solver/blob/master/src/main/java/org/chocosolver/solver/Solver.java#L258)
When this method detects inconsistency, it [restores the domain to their initial value](https://github.com/chocoteam/choco-solver/blob/master/src/main/java/org/chocosolver/solver/Solver.java#L365) and also makes sure there is [no pending event in the propagation engine](https://github.com/chocoteam/choco-solver/blob/master/src/main/java/org/chocosolver/solver/Solver.java#L360).
Then, `solve()` method returns `false` and you have to check the number of nodes (set to 0 if initialize fails).

Then, about your variants:
1. why not, but this is redundant with nb nodes = 0 ;
2. this is already done in `closeSearch()` which is always called after `initialize()` in `solve()` ;
3. Since `initialize()` was called, there is no reason to not call `afterInitialize()`.

Best,

**Comment by gregy4** [comment]
In my scheduling problem are some relations between tasks very strict and often chained. To not repeat deduction work done by solver already I identify groups of tasks with strict relations and based on domains of tasks variables in these groups is deduced required minimal number of resources. In case number of available resources is lower computation is finished with a reason not enough resources.

When initial propagation fails, and tasks domains are not reduced, my code in the method can deduce that there’s no enough resources even with unlimited time but real reason can be that shortest possible schedule is longer than given boundaries. Small but very important difference :-)

To summarize, when a code inside afterInitialize method analyze reduced domains to stop solver or reduce search space it is definitely benefical to know that initial propagation already failed.

Based on your response variant 1 seems to me like the best one.

**Comment by cprudhom** [comment]
Ok, so you can check the nb of failures too in `afterInitialize()` ? If equal to 1, then initial propagation failed. Does that fit your need ?

**Comment by gregy4** [comment]
Yes, it fits my needs. I wrote about it in original post (Measure.failCount). For choco 3.3.3 it is good solution, for current version is probably better extra argument to afterInitialize method.
- other users can realize that there are two cases for afterInitialize method, one for succesful and one for not succesful initial propagation and can benefit from this information
- new argument to afterInitialize method protects against future changes in which nb of failures > 0 can mean something different to fail in initial propagation
